### ISSUE TITLE
Detach Links without Closing

### ISSUE THREAD

**Issue by PiotrSpikowski** [Azure/amqpnetlite#213]
Currently it is possible to fully close a link. A call to the ``Close`` method will always send a ``Detach`` performative with a ``closed`` field set to true.

We need to also have a possibility to detach a link without closing it (``closed`` set to ``false``). On the ``ReceiverLink`` one may set a terminus expiriation policy (link-detach) together with a timeout value. In such a case the expiration timer starts when a terminus is detached. We want a durable subscription (with remaining ttl) to continue for the given period of time. For this, a Detach performative with a closed set to false is needed.

The current implementation is not opened for extensions so it is not possible to achieve the above goal.

If the ``Close`` method is always setting close to ``true`` we can image a new one e.g. simply called ``Detach`` and setting the closed flag to ``false``.

```cs
public interface ILink 
{
       
        ...

        /// <summary>
        /// Detaches the link sending a detach performative and altering the link state.
        /// </summary>
        /// <param name="error">An error causing a detach.</param>
        bool Detach(Error error);

        /// <summary>
        /// Detaches the link sending a detach performative and altering the link state.
        /// </summary>
        bool Detach();

        ...

}
```
The above is showing what could be added. 

Regards
Piotr

**Comment by xinchen10** [comment]
@PiotrSpikowski do you have any suggestion on what to do with the link after it is detached. With the current implementation, the link object is unusable after close/detach. The application would need to create a new link object with the same name to resume. This should be fine except that the unsettled map is not accessible after a link is detached, and that is required to do appropriate link resume.

**Comment by PiotrSpikowski** [comment]
@xinchen10 I need to mainly detach a link without closing it on the receiving side (``ReceiverLink``). This is in order to ensure that a durable subscription is not cancelled. 

When a receiving link is detached it might be that we have not sent all the dispositions to the already received messages (at least in theory). In such a case a broker will however “release” all the messages on their side and transfer them to other active subscribers. In other words we do not have to worry too much about the state of a detached, receiving link. At least this is how I can see it.

It might be a bit different if we are acting as a sending side. We send a message and wait for a disposition. Theoretically if such a link is detached we should keep the unsettle map so we may for example repeat the transfer. The current implementation is however not supporting reattachment of links (e.g. an attach frame is sent in the link ctor). As resuming is not supported I can see two options:
-	We add the Detach option to the receiving link only
-	We add it to both but we agree that there is no possibility to reattach right now anyway. One must re-create the link with all the consequences which for me is acceptable

**Comment by xinchen10** [comment]
@PiotrSpikowski that sounds good. We will review the PR after you complete the CLA.

**Comment by cjwmorgan-sol-sys** [comment]
Hi I'm interested in this I as face the same issue to ensure a durable subscription is not cancelled. Any idea how long until this issue can be fixed?

**Comment by xinchen10** [comment]
Available in 2.1.0.
